File: centum_flutter/lib/presentation/components/category_dialog.dart
import 'package:flutter/material.dart';
import 'package:centum/core/theme/app_theme.dart';
import 'package:centum/core/constants/dimensions.dart';
import 'package:centum/data/repositories/category_repository.dart';

class CategoryDialog extends StatefulWidget {
  const CategoryDialog({super.key});

  @override
  State<CategoryDialog> createState() => _CategoryDialogState();
}

class _CategoryDialogState extends State<CategoryDialog> {
  final _nameController = TextEditingController();
  String _selectedType = 'expense';
  String? _selectedColor;

  final _colors = [
    '#4CAF50', '#EF5350', '#42A5F5', '#AB47BC',
    '#FF7043', '#26C6DA', '#FFEE58', '#66BB6A',
    '#EC407A', '#78909C',
  ];

  @override
  void dispose() {
    _nameController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      backgroundColor: AppColors.surface,
      title: const Text('Новая категория',
          style: TextStyle(color: AppColors.text)),
      content: SizedBox(
        width: double.maxFinite,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            SegmentedButton<String>(
              segments: const [
                ButtonSegment(value: 'income', label: Text('Доход')),
                ButtonSegment(value: 'expense', label: Text('Расход')),
              ],
              selected: {_selectedType},
              onSelectionChanged: (v) =>
                  setState(() => _selectedType = v.first),
              style: ButtonStyle(
                backgroundColor: WidgetStateProperty.resolveWith((states) {
                  if (states.contains(WidgetState.selected)) {
                    return _selectedType == 'income'
                        ? AppColors.income.withOpacity(0.2)
                        : AppColors.expense.withOpacity(0.2);
                  }
                  return AppColors.card;
                }),
              ),
            ),
            const SizedBox(height: Dimensions.md),
            TextField(
              controller: _nameController,
              decoration: const InputDecoration(labelText: 'Название'),
            ),
            const SizedBox(height: Dimensions.md),
            const Text('Цвет', style: TextStyle(color: AppColors.subtext)),
            const SizedBox(height: Dimensions.sm),
            Wrap(
              spacing: Dimensions.sm,
              runSpacing: Dimensions.sm,
              children: _colors.map((c) {
                final color = Color(int.parse(c.replaceFirst('#', '0xFF')));
                final isSelected = _selectedColor == c;
                return GestureDetector(
                  onTap: () => setState(() => _selectedColor = c),
                  child: AnimatedContainer(
                    duration: const Duration(milliseconds: 200),
                    width: 32,
                    height: 32,
                    decoration: BoxDecoration(
                      color: color,
                      borderRadius: BorderRadius.circular(8),
                      border: isSelected
                          ? Border.all(color: AppColors.text, width: 2)
                          : null,
                    ),
                  ),
                );
              }).toList(),
            ),
          ],
        ),
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.pop(context),
          child: const Text('Отмена',
              style: TextStyle(color: AppColors.subtext)),
        ),
        ElevatedButton(
          onPressed: _submit,
          style: ElevatedButton.styleFrom(
            backgroundColor: AppColors.accent,
          ),
          child: const Text('Создать',
              style: TextStyle(color: AppColors.text)),
        ),
      ],
    );
  }

  Future<void> _submit() async {
    final name = _nameController.text.trim();
    if (name.length < 2) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Минимум 2 символа')),
      );
      return;
    }
    try {
      await CategoryRepository.createCategory(name, _selectedType, _selectedColor);
      if (mounted) Navigator.pop(context, true);
    } catch (e) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text('Ошибка: $e')),
        );
      }
    }
  }
}

File: Centum/views/components/category_dialog.py
import flet as ft
from utils.theme import C, FS, SP, R, tf_style
from models.category import create_category

COLORS=["#4CAF50","#EF5350","#42A5F5","#AB47BC",
        "#FF7043","#26C6DA","#FFCA28","#66BB6A","#EC407A","#78909C"]


class CategoryDialog(ft.AlertDialog):
    def __init__(self, on_success=None):
        self._cb=on_success; self._color=COLORS[0]; self._type="expense"
        self._name=ft.TextField(label="Название категории", max_length=40, **tf_style())
        self._seg=ft.SegmentedButton(
            selected=["expense"],
            segments=[
                ft.Segment(value="income",  label=ft.Text("Доход",  color=C.TEXT)),
                ft.Segment(value="expense", label=ft.Text("Расход", color=C.TEXT)),
            ],
            on_change=lambda e: setattr(self, '_type', list(e.control.selected)[0]),
        )
        self._crow = ft.Row(controls=self._chips(), wrap=True, spacing=SP.SM)
        self._err  = ft.Text("", color=C.EXPENSE, size=FS.SM, visible=False)

        super().__init__(
            modal=True, bgcolor=C.CARD,
            shape=ft.RoundedRectangleBorder(radius=R.MD),
            title=ft.Text("Новая категория", color=C.TEXT, size=FS.LG, weight=ft.FontWeight.BOLD),
            content=ft.Column(controls=[
                self._seg, self._name,
                ft.Text("Цвет метки", color=C.SUBTEXT, size=FS.SM),
                self._crow, self._err,
            ], spacing=SP.SM, tight=True, width=300),
            actions=[
                ft.FilledButton(
                    content=ft.Text("Отмена", color=C.SUBTEXT, size=FS.MD),
                    style=ft.ButtonStyle(
                        bgcolor={ft.ControlState.DEFAULT: C.SURFACE},
                        shape=ft.RoundedRectangleBorder(radius=R.SM)),
                    on_click=lambda e: e.page.pop_dialog(),
                ),
                ft.FilledButton(
                    content=ft.Text("Создать", color=C.BG, size=FS.MD, weight=ft.FontWeight.BOLD),
                    style=ft.ButtonStyle(
                        bgcolor={ft.ControlState.DEFAULT: C.ACCENT},
                        color={ft.ControlState.DEFAULT: C.BG},
                        shape=ft.RoundedRectangleBorder(radius=R.SM)),
                    on_click=self._confirm,
                ),
            ],
            actions_alignment=ft.MainAxisAlignment.END,
        )

    def open_dialog(self, page, default_type="expense"):
        self._type=default_type; self._seg.selected=[default_type]
        self._name.value=""; self._err.visible=False
        page.show_dialog(self)

    def _chips(self):
        return [
            ft.GestureDetector(
                content=ft.Container(
                    width=30, height=30, bgcolor=c, border_radius=R.FULL,
                    border=ft.border.all(3, C.TEXT if c==self._color else "transparent"),
                    animate=ft.Animation(150, ft.AnimationCurve.EASE_IN_OUT),
                ),
                on_tap=lambda e, col=c: self._pick(col),
            ) for c in COLORS
        ]

    def _pick(self, color):
        self._color=color; self._crow.controls=self._chips(); self._crow.update()

    def _confirm(self, e):
        name=(self._name.value or "").strip()
        if len(name)<2:
            self._err.value="Мин. 2 символа"; self._err.visible=True; self.update(); return
        try:
            cat=create_category(name=name, cat_type=self._type, color=self._color)
            e.page.pop_dialog()
            if self._cb: self._cb(cat)
        except Exception as ex:
            self._err.value=f"Ошибка: {ex}"; self._err.visible=True; self.update()
